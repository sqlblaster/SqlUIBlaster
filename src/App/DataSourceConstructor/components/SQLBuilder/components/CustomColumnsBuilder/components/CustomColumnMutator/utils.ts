import { ColumnBranch } from '../../../common/ColumnSelector/ColumnItem/SelectedColumn.models';
import { CustomColumn } from '../CustomColumnItem/model';

export const filterCustomColumn = (
  customColumnBranch: ColumnBranch,
  customColumns: CustomColumn[]
): boolean =>
  !customColumns.find(({ column }) => column.equals(customColumnBranch)) &&
  customColumnBranch !== customColumnBranch.lastColumn;
