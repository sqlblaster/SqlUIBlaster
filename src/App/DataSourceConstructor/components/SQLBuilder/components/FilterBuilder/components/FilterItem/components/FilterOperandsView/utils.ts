import { Operation } from '../../../MutateFilterPopover/Operation';
import { DateOperation } from '../../../MutateFilterPopover/SwitchMutator/mutators/DateMutator/DateOperation';
import { CoupleOperandsViewProps } from './components/CoupleOperandsView';
import { SingleOperandViewProps } from './components/SingleOperandView';
import {
  formatCoupleDateOperands,
  formatNearDateOperands,
  formatSingleDateOperands
} from './date-formatter';

export const hasOperands = (operator: Operation['operator']): boolean => {
  switch (operator) {
    case 'Is empty':
    case 'Is not empty':
    case 'Is true':
    case 'Is false':
      return false;
    default:
      return true;
  }
};

export const getFormattedDateOperand = (operation: DateOperation): string => {
  switch (operation.operator) {
    case 'Previous':
    case 'Next':
      return formatNearDateOperands(operation.operands, operation.operator);
    case 'Before':
    case 'After':
    case 'On':
      return formatSingleDateOperands(operation.operands);
    default:
      return '';
  }
};

export const getFormattedCoupleOperands = (
  operation: Operation
): CoupleOperandsViewProps['operands'] => {
  if (operation.type === 'Date' && operation.operands instanceof Array) {
    return formatCoupleDateOperands(operation.operands);
  } else if (
    operation.type === 'Number' &&
    operation.operands instanceof Array
  ) {
    return operation.operands;
  } else {
    throw new Error('expected between operands');
  }
};

export const getFormattedSingleOperand = (
  operation: Operation
): SingleOperandViewProps['operand'] => {
  if (
    hasOperands(operation.operator) &&
    !(operation.operands instanceof Array) &&
    hasOperands(operation.operator)
  ) {
    switch (operation.type) {
      case 'Foreign': {
        const record = operation.operands;

        return record && record.title ? record.title : '';
      }
      case 'Date':
        return getFormattedDateOperand(operation);
      case 'Number':
      case 'String':
      default:
        return operation.operands as number | string;
    }
  } else {
    throw new Error('expected singular operand');
  }
};
