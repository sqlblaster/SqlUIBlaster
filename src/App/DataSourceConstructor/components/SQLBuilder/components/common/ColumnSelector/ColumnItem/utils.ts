import { clone } from 'ramda';
import { ColumnBranch } from './SelectedColumn.models';

export const getNewTableBranch = (
  parentColumnBranch: ColumnBranch,
  foreignTableName: ColumnBranch['foreignTableName'],
  foreignModelName: ColumnBranch['foreignModelName']
) => {
  const newTableBranch = clone(parentColumnBranch);

  newTableBranch.lastColumn.foreignColumn = new ColumnBranch({
    tableName: foreignTableName,
    modelName: foreignModelName
  });

  newTableBranch.lastColumn = newTableBranch.lastColumn.foreignColumn;

  return newTableBranch;
};

export const listItemOwnPaddingLeft: number = 16;
export const paddingShiftAmountByLevel: number = 15;
export const calcPaddingAccordingTo = (level: number) => {
  return listItemOwnPaddingLeft + level * paddingShiftAmountByLevel;
};
