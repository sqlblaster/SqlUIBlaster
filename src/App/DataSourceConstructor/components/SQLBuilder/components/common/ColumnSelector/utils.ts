import { clone } from 'ramda';
import { Column } from 'src/App/DataSourceConstructor/schemas';
import { ColumnBranch } from './ColumnItem/SelectedColumn.models';
import { ColumnSelectorOwnProps } from './props';

export const getRootTableBranch = (tableName: string, modelName: string) =>
  new ColumnBranch({ tableName, modelName });

export const getNewColumnBranch = (
  tableBranch: ColumnBranch,
  { name, type, foreignTableName, foreignModelName }: Column
) => {
  const newColumnBranch = new ColumnBranch(clone(tableBranch));

  const lastColumn = newColumnBranch.lastColumn;
  lastColumn.columnName = name;
  lastColumn.type = type;
  lastColumn.foreignTableName = foreignTableName;
  lastColumn.foreignModelName = foreignModelName;

  return newColumnBranch;
};

export const sortInOrderForeignColumnsBelow = (columns: Column[]) =>
  columns.sort((c1, c2) => +!!c1.foreignTableName - +!!c2.foreignTableName);

export const filterColumns = (
  columns: Column[],
  canShowColumnBranch: Required<ColumnSelectorOwnProps>['canShowColumnBranch'],
  tableBranch: ColumnBranch
): Column[] => {
  return columns.filter(
    column =>
      canShowColumnBranch(getNewColumnBranch(tableBranch, column)) ||
      column.foreignTableName
  );
};
