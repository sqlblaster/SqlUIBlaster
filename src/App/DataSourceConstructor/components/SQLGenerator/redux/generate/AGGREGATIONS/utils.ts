import { ColumnBranchAliased } from '../JOIN/models/ColumnAliased';

export const getAlias = (
  columnBranch: ColumnBranchAliased | undefined,
  operator: string
): string => {
  let alias = operator;

  if (columnBranch) {
    const {
      lastColumn: { tableAlias, columnName }
    } = columnBranch;

    alias += ` ${
      columnBranch.hasZeroColumnLevel() ? columnName : `${columnName} (${tableAlias})`}`;
  }

  return alias;
};
