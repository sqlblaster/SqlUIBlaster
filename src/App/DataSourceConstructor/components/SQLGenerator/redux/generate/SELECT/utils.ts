import { ColumnBranchAliased } from '../JOIN/models/ColumnAliased';

export const getColumnAlias = (columnBranch: ColumnBranchAliased): string => {
  const {
    lastColumn: { columnName, tableAlias }
  } = columnBranch;

  return `AS "${columnName} (${tableAlias})"`;
};
