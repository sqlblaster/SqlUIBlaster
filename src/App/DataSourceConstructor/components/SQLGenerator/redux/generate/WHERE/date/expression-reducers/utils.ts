import * as moment from 'moment';
import { PluralDateComponent } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/FilterBuilder/components/MutateFilterPopover/SwitchMutator/mutators/DateMutator/SwitchDateOperands/DateOperandsComponents/NearDate/plural-date-components';
import { RelativeDate } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/FilterBuilder/components/MutateFilterPopover/SwitchMutator/mutators/DateMutator/SwitchDateOperands/DatePickerByOperandsType/components/RelativeDatePicker/singular-date-components';

export type Format = 'date' | 'timestamp';

export const DateComponentToFormatMap: {
  [K in PluralDateComponent]: string
} = {
  Years: 'YYYY',
  Months: 'MM',
  Weeks: 'WW',
  Days: 'DD',
  Hours: 'HH24',
  Minutes: 'MI',
  Seconds: 'SS'
};

export const truncDate = (
  dateComponent: PluralDateComponent,
  date: Date | string
): string => {
  return `date_trunc('${dateComponent}', (${
    date instanceof Date ? toSqlDate(date) : date
  })::timestamp)`;
};

export const toSqlDate = (date: Date | string = 'now()') => {
  return date instanceof Date
    ? `'${moment(date).format('YYYY-MM-DD HH:mm:ss')}'`
    : date;
};

export type DateParam = Parameters<typeof toSqlDate>[0];

export const addDateSql = (
  shiftAmount: number,
  dateComponent: PluralDateComponent,
  date: DateParam = 'now()'
): string => {
  return `${toSqlDate(date)} + INTERVAL '${shiftAmount} ${dateComponent}'`;
};

export const getRelativeDateSqlExp = (relativeDate: RelativeDate): string => {
  switch (relativeDate) {
    case 'Today':
      return `now()::date`;
    case 'Beginning of the current week':
      return `date_trunc('week', now())`;
    case 'End of the current week':
      return `(
        date_trunc('week', now()) + interval '6 days'
      )`;
    case 'Beginning of the current month':
      return `date_trunc('month', now())`;
    case 'End of the current month':
      return `(
        date_trunc('month', now()) + interval '1 month - 1 day'
      )`;
    case 'Beginning of the current quarter':
      return `date_trunc('quarter', now())`;
    case 'End of the current quarter':
      return `(
        date_trunc('quarter', now()) + interval '3 month - 1 day'
      )`;
    case 'Beginning of the current year':
      return `date_trunc('year', now())`;
    case 'End of the current year':
      return `(
        date_trunc('year', now()) + interval '1 year - 1 day'
      )`;
    default:
      throw new Error('invalid relative date');
  }
};
