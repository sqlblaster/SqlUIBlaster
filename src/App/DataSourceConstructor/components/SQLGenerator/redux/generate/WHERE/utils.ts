import { formatColumn } from '../common-formatters';
import { ColumnBranchAliased } from '../JOIN/models/ColumnAliased';

export const getColumnString = ({
  tableAlias,
  columnName
}: ColumnBranchAliased): string => formatColumn(columnName, tableAlias);
