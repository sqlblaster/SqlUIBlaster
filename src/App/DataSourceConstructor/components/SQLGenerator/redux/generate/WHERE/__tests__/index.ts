import { generateWhere } from '..';
import { complexBracketsFilters } from './utils/complex-brackets-filters';
import { doubleBracketsFilter } from './utils/double-brackets-filter';
import { singleBracketsFilter } from './utils/single-brackets-filter';

describe('WHERE clause', () => {
  it('should render nothing when there are no filters', () => {
    expect(generateWhere([])).toMatchSnapshot();
  });

  it('should render open bracket and closing bracket', () => {
    expect(generateWhere(singleBracketsFilter)).toMatchSnapshot();
  });

  it('should render both open and closing brackets', () => {
    expect(generateWhere(doubleBracketsFilter)).toMatchSnapshot();
  });

  it('should render complex brackets', () => {
    expect(generateWhere(complexBracketsFilters)).toMatchSnapshot();
  });
});
