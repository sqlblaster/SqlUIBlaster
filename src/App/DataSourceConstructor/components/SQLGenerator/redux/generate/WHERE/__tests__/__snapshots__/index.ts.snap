// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WHERE clause should render both open and closing brackets 1`] = `
"
    WHERE  ((\\"orders1\\".\\"Details\\"::text = 'Detail1'))"
`;

exports[`WHERE clause should render complex brackets 1`] = `
"
    WHERE  ((\\"orders1\\".\\"Details\\"::text = 'Detail1' or \\"orders1\\".\\"Details\\"::text = 'Detail2') and \\"orders1\\".\\"Quantity\\" > 10) and (\\"orders1\\".\\"Details\\"::text = 'Detail1' and (\\"orders1\\".\\"Quantity\\" < 5 or \\"orders1\\".\\"Quantity\\" IS NULL))"
`;

exports[`WHERE clause should render nothing when there are no filters 1`] = `""`;

exports[`WHERE clause should render open bracket and closing bracket 1`] = `
"
    WHERE  (\\"orders1\\".\\"Details\\"::text = 'Detail1')"
`;
